src/support/hash.h

```diff
@@ -45,6 +45,11 @@ inline uint64_t rehash(uint64_t x, uint64_t y) {
   return rehash(ret, HashType(y >> 32));
 }
 
+template<typename T> inline void hash_combine(std::size_t& s, const T& v) {
+  // see boost/container_hash/hash.hpp
+  s ^= std::hash<T>{}(v) + 0x9e3779b9 + (s << 6) + (s >> 2);
+}
+
```

Comment:
This looks like a nice general replacement for `rehash`. Could we continue using `rehash` for now in the PR and split out a separate PR to replace `rehash` with this `hash_combine` everywhere?

Reply:
Introduced this one because it was easier for me to reason about since it is hashing exactly the type of value I provide by using `std::hash`, incl. bools which `rehash` doesn't cover explicitly (probably because it doesn't had to so far; can probably cast to `uint32_t` but that seems dangerous because it hashes a lot of zeroes every time). Plus, the differentiation between `uint32_t` and `uint64_t` and sometimes providing a `0` as a first argument looked like I'd be very easy to use `rehash` wrong.

I do understand however that it makes sense to separate it from this PR, but it is very likely that the code will be somewhat I don't even know how to call it then with me being unable to reason about what could go wrong. But I can try if you wish?

Reply:
We've been using `rehash` in a very unscientific way: just cast everything that needs to be hashed to either `uint32_t` or `uint64_t` and throw it in. Sometimes we pack multiple fields into a single `uint32_t` or `uint64_t` if possible but we mostly don't worry about it. The bar for using `rehash` "correctly" is very very low. That's why broader hashing improvements would be welcome.

Now that I think about it, though, I think I remember that @kripken wanted to avoid using `size_t` in the hashing to avoid having differences between 32-bit and 64-bit platforms. I'm not sure what his latest thinking on that is, though.

Reply:
Ah, so this is about being deterministic between archs? If so, then this problem wasn't completely avoided before already due to

https://github.com/WebAssembly/binaryen/blob/9c84d90712c1183f1d9ab3e5068d902eb7d627ea/src/wasm/wasm-type.cpp#L41-L43

(code before this PR) returning a `size_t`. Not sure what exactly this tells us, though.

Reply:
If I do remember correctly that being deterministic was the original goal, then I believe it is also true that we no longer care so much. I would personally be happy to use size_t everywhere, but I think it does make more sense as a separate change.

Reply:
Currently looking into changing everything to `wasm::rehash`. What I'm wondering, besides how this is deterministic in the first place, is how

```cpp
size_t hash<wasm::Signature>::operator()(const wasm::Signature& sig) const {
  return wasm::rehash(uint64_t(hash<uint64_t>{}(sig.params.getID())),
                      uint64_t(hash<uint64_t>{}(sig.results.getID())));
}
```

(this is one of the previous functions) does compile on 32-bit platforms, considering that `wasm::rehash(uint64_t, uint64_t)` returns `uint64_t`, so wouldn't that attempt to do an implicit conversion from `uint64_t` to `size_t`? Or is there some magic pinning `size_t` to 64-bit on 32-bit platforms? Not arguing against doing the change, as I can probably do it this way everywhere, just not really comfortable with it.

Reply:
Created a [PR to revert and adapt to `wasm::rehash` on top of this PR](https://github.com/dcodeIO/binaryen/pull/2), commenting on some of the aspects I'm not comfortable with. Hope this helps to decide how to proceed, with just merging it into here and not worrying about it too much as suggested being one option. Alternatively, I could look into replacing `wasm::rehash` with the alternative `wasm::hash_combine` as a preliminary PR to this one, but that would imply that determinism between 32- and 64-bit builds isn't guaranteed for sure, which I guess currently isn't and cannot be achieved anyway because ultimately `std::unordered_map` etc. are going to use `size_t`, no matter if we like it or not, and because we are hashing `uintptr_t`s as `uint64_t`s when dealing with constructed types.

Reply:
Superseded by https://github.com/WebAssembly/binaryen/pull/3023 :)